staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +35,498 @@ func (s *Source) ServiceNodeDelete(path string) {
-	scope.Infof("try to watch instance under the zk path: %q ", path)
-	initWatch := true
 	for {
-		pathChildren, _, pathEvent, err := s.Con.ChildrenW(path)
+		select {
+		case <-ew.exit:
+			return
+		default:
+		}
+
+		paths, _, eventCh, err := ew.conn.Load().(*zk.Conn).ChildrenW(path)
 		if err != nil {
-			if shouldStopWatch(err) {
-				if !initWatch {
-					if deleteFunction != nil {
-						deleteFunction(path)
-					}
-					s.watchPath.Remove(path)
-				}
-				scope.Errorf("stop watch instance under the zk path %q with err: %v", path, err)
-				return
-			}
-			scope.Warnf("watch instance under the zk path %q failed: %s", path, err)
-			time.Sleep(time.Second)
+			time.Sleep(b.Duration())
+			continue
 		}
-		if initWatch {
-			s.watchPath.Set(path, nil)
-			scope.Infof("watching instance under the zk path: %q", path)
-			initWatch = false
+		revision := resultRevision()
+		b.Reset()
+
+		select {
+		case ch <- watchResultWrapper{
+			paths:    paths,
+			revision: revision,
+		}:
+		default:
 		}
-		provider := false
-		providersPath := path[0:len(path)-10] + "/" + ProviderNode
-		consumersPath := path
-		if strings.HasSuffix(path, ProviderNode) {
-			provider = true
-			providersPath = path
-			consumersPath = path[0:len(path)-10] + "/" + ConsumerNode
+		<-eventCh
+	}
+}
+
+func (ew *EndpointWatcher) watchBoth(ctx context.Context) {
+	defer close(ew.exit)
+
+	// Try to initialize, but it is not required to be completed,
+	// because the service may have been deleted or for others.
+	providers, _, err := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + providerPathSuffix)
+	if err == nil {
+		// ignore errors when getting consumer fails
+		consumers, _, _ := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + consumerPathSuffix)
+		ew.handler(providers, consumers)
+	}
+	ew.initCallback()
+
+	providerEventCh, consumerEventCh := make(chan watchResultWrapper), make(chan watchResultWrapper)
+	go ew.simpleWatch(ew.servicePath+providerPathSuffix, providerEventCh)
+	go ew.simpleWatch(ew.servicePath+consumerPathSuffix, consumerEventCh)
+	var providersRev, consumersRev int64
+	for {
+		// Delete event highest priority
+		select {
+		case <-ew.signalExit:
+			ew.serviceDeleteHandler()
+			scope.Infof("endpointWatcher %q exit due to service deleted", ew.servicePath)
+			return
+		default:
 		}
-		if provider {
-			otherChild, _, err := s.Con.Children(consumersPath)
-			if err != nil {
-				scope.Warnf("get children under the zk path %q failed: %s", consumersPath, err)
-			} else {
-				updateFunction(pathChildren, otherChild, path)
+
+		select {
+		case <-ctx.Done():
+			scope.Debugf("endpointWatcher %q exit due to ctx.Done()", ew.servicePath)
+			return
+		default:
+		}
+
+		select {
+		case providerWatchResult := <-providerEventCh:
+			if providerWatchResult.revision < providersRev {
+				continue
+			}
+			providersRev = providerWatchResult.revision
+			consumers, _, err := ew.conn.Load().(*zk.Conn).Children(ew.servicePath + consumerPathSuffix)
```

Comment:
the `simpleWatch` actually transfer the data but we do another fetch. Add a `TODO` for later optimization